DOC Update writing style guide (#494)
silverstripe/developer-docs: en/10_Contributing/02_Documentation.md

Reviewer: The following points are being removed without explanation:

1. Sentence case your bullet points. If a bullet point is a full sentence then end with a full stop. If it is a sentence fragment or a comma separated list, full stops are not required.
1. We use sentence case for titles. Only capitalise the first letter of the first word of a title. The only exceptions to this are when using brand names (e.g. GitHub), acronyms (e.g. PHP) and class names (e.g. `ModelAdmin`).
1. Use an empty line to separate paragraphs.
1. Never leave spaces at the end of a line.
1. Avoid saying words like "obviously" or "of course". Things that are obvious to you as the writer may not be so obvious to the person reading the documentation.
1. If referring to maintainers of Silverstripe CMS, either explicitly say "the maintainers", or write in second person plural form (use "we" instead of "I").

I've made them a numbered list so we can refer to them easily.

I'm okay with dropping 1 because I don't know if we even follow that right now.
I'm also okay with dropping 3 and 4 because I'm 90% sure the linter will pick those up anyway.
The rest I think should be kept.

Author: Have added in points 2, 4 and 6

Reviewer: You added in points 2, 5, and 6 which I assume is what you meant - it's also what I wanted so 👍

## Writing style

> - Write in second person form, addressing the reader. For example "First you'll install a webserver" is good style.

Author: I'm personally not a fan of the "Write in second person form" because IMO it  just adds unnecessary fluff. I much prefer dry technical documentation that simply tells you what to do i.e. "First install a webserver".

Reviewer 2: Also prefer direct and instructional approach rather than a story.

Reviewer: I think that one was more intended to prefer "you can configure this by..." instead of "this can be configured by..." which is just another way of saying "use an active voice".

- Use an active voice.
- Use plain language for easier comprehension.
- Keep sentences clear and concise.

### Content structure
- Start each page with a clear summary sentence outlining what readers can expect.
- Structure content around a single topic.
- Use appropriate markdown. The use of code blocks, lists, tables and callout blocks make technical writing easier to scan.

> - Use descriptive headings and subheadings organised in a hierarchy

Reviewer 2: ```suggestion
- Use descriptive headings and subheadings organised in a hierarchy.
```

Author: Done

- Present lists as bullet points.
- Avoid FAQs as they are not a replacement for coherent, well explained documentation.

### Language and grammar

> - Use UK English, adhering to -ize vs -ise, and -or vs our spelling conventions.

Reviewer 2: ```suggestion
- Use UK English, adhering to -ize vs -ise, and -or vs -our spelling conventions.
```

Author: Done

- Spell out acronyms the first time, with the abbreviation in brackets.
- Spell out numbers at the beginning of sentences; otherwise, use numerals.
- Use simple contractions for readability.

> - Use the serial (Oxford) comma.

Reviewer 2: ```suggestion
```

Author: Removed

- Only use ampersands for brand or team names.
- Use gender-neutral language and singular "they" instead of "he/she/his/her".
- Use "allowlist/denylist" instead of "whitelist/blacklist".
- Refrain from underlining, italicizing, bolding, or using all caps. Instead use callout blocks where appropriate.

> - Use meaningful link text, i.e. avoid the use of "click here" link text.

Reviewer 2: ```suggestion
- Use meaningful link text, i.e. avoid the use of "click here" link text.
- Use i.e. and e.g. with dots.
```

Author: Added

### Code blocks
- Use triple backtick code fences to denote code blocks.
- All code fences must define a language. Use `text` for plain text.
- Avoid using a $ symbol for bash code blocks.
- Code examples must follow [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/) for that language.

> - Use "App" as the PHP namespace when defining code that meant to be added at the project level.

Reviewer 2: ```suggestion
- Use "App" as the PHP namespace when defining code that is meant to be added at the project level.
```

Author: Done

- Include the filename that a code block relates to, do so in a comment in the first line of the code block. Don't place a blank line after this comment, unless the next line is also a comment.

### References to various things
- Add a trailing slash to directory references to distinguish from regular files.
- [Link to Silverstripe API docs](#links-to-api-documentation) for the first reference to a Silverstripe PHP class on a page. After that, surrond the class name with backticks.
- Use `example.com` as the domain of sample URLs and `example.org` and `example.net` when additional domains are required. All of these domains are [reserved by the IANA](https://tools.ietf.org/html/rfc2606#section-3).

### References to Silverstripe

> - Spell "Silverstripe" with one uppercase S and one lowercase s.

Reviewer 2: I guess we don't want silverStripe
```suggestion
- Spell "Silverstripe" with an uppercase 'S' at the beginning and a lowercase 's' in the middle.
```

Author: Done

> - Refer to the company as "Silverstripe" or "Silverstripe Ltd."

Reviewer 2: I'd rather if ambiguous Silverstripe meant the CMS, not the company, but that's probably not gonna fly :-))
```suggestion
- Refer to the company as "Silverstripe Ltd."
```

Author: Done

- Use "Silverstripe CMS" to refer the content management system.

### Markdown linting

> - Files that are within a directory speicied in a repositories `.doclintrc` file will be automatically linted in CI using [silverstripe/documentation-lint](https://github.com/silverstripe/documentation-lint/).

Reviewer 2: ```suggestion
- Files that are within a directory specified in repository's `.doclintrc` file will be automatically linted in CI using [silverstripe/documentation-lint](https://github.com/silverstripe/documentation-lint/).
```

Author: Done

## Writing style

### Voice and tone

- Use an active voice.

> - Use plain language for easier comprehension.

Reviewer: ```suggestion
- Use plain language for easier comprehension. Avoid uncommon jargon and overly long words. Remember that not everyone speaks English as their primary language.
```
I think this extra detail from the existing docs is useful

Author: Done

- Keep sentences clear and concise.

### Content structure

> - Start each page with a clear summary sentence outlining what readers can expect.

Reviewer: ```suggestion
```
That's a good goal but our current docs don't follow that rule - unless we intend to add this to every page as part of this PR we should probably just not say this.

Author: Done

- Structure content around a single topic.
- Use appropriate markdown. The use of code blocks, lists, tables and callout blocks make technical writing easier to scan.
- Use descriptive headings and subheadings organised in a hierarchy.

> - Present lists as bullet points.

Reviewer: ```suggestion
```
Lists can be ordered, or use `<dl>` - not all lists should be unordered bulleted lists.

If the intention is instead to avoid comma separating items in a sentence when they should be a semantic list, we should specify that more directly.

Author: Done

- Avoid FAQs as they are not a replacement for coherent, well explained documentation.

### Language and grammar

> - Use UK English, adhering to -ize vs -ise, and -or vs -our spelling conventions.

Reviewer: ```suggestion
- Use UK English. The most common differences are -ize vs -ise, and -or vs -our (eg color vs colour).
```
I think the way the original said it was clearer - so this brings some of that back.

Author: Done

> - Spell out acronyms the first time, with the abbreviation in brackets.

Reviewer: ```suggestion
- Spell out acronyms the first time, with the abbreviation in brackets.
    - If an acronym is well established, such as "API", you don't need to spell it out.
```

Author: Done

> - Spell out numbers at the beginning of sentences; otherwise, use numerals.

Reviewer: ```suggestion
```
I don't think we need a rule for this - it's unlikely any of us will remember to do it.

Author: Done

> - Use simple contractions for readability.

Reviewer: ```suggestion
```
I don't think we need a rule for this

Author: Done

> - Only use ampersands for brand or team names.

Reviewer: ```suggestion
```
I don't know what this even means - but we don't have any team names or brands that need this rule.

Author: Done

- Use gender-neutral language and singular "they" instead of "he/she/his/her".

> - Use "allowlist/denylist" instead of "whitelist/blacklist".

Reviewer: ```suggestion
```
This would need to be accompanied with a host of code changes to be viable

Author: Done

> - Refrain from underlining, italicizing, bolding, or using all caps. Instead use callout blocks where appropriate.

Reviewer: ```suggestion
```
This doesn't seem sensible to me. A callout block is not a substitute for appropriate emphasis in the middle of a sentence.

Author: Done

- Use meaningful link text, i.e. avoid the use of "click here" link text.

> - Use i.e. and e.g. with dots.

Reviewer: I prefer "for example" over "e.g." - what is the intention here @michalkleiner?

Author: I think he means don't use "eg", use "e.g." instead

We have a million matches on "e.g." right now in the docs, shouldn't changes these IMO

Reviewer: That's okay then. If someone in the future tells me to change "for example" to "e.g." I can send them here and tell them no. 👌
Totally agree that `eg` and `ie` are wrong though.

### Code blocks

> - Use triple backtick code fences to denote code blocks.

Reviewer: ```suggestion
```
Doesn't need to be said

Author: Done

> - All code fences must define a language. Use `text` for plain text.

Reviewer: ```suggestion
- All code fences must have an appropriate syntax highlighting hint. Use `text` for plain text.
```
I think this was clearer from the original

Author: Done

> - Avoid using a $ symbol for bash code blocks.

Reviewer: ```suggestion
- Do not put a `$` in front of lines of bash code examples.
```
Not avoid it - just don't do it ever.

Author: Done

> - Code examples must follow [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/) for that language.

Reviewer: ```suggestion
- Code examples must follow [coding conventions](/contributing/coding_conventions/) for that language.
```

Author: Done

> - Use "App" as the PHP namespace when defining code that is meant to be added at the project level.

Reviewer: ```suggestion
- Use `App` when the code requires a vendor name (e.g. as the root for namespaces).
```

It's not just PHP namespaces, that's just the most common scenario.

Author: Done

> - Include the filename that a code block relates to, do so in a comment in the first line of the code block. Don't place a blank line after this comment, unless the next line is also a comment.

Reviewer: ```suggestion
- If it's useful to include the filename that a code block relates to, do so in a comment in the first line of the code block.
  - Don't place a blank line after this comment, unless the next line is also a comment.
```
The original was better imo. A lot of the time a filename doesn't provide any useful context, so it's not worth adding.

Author: Done

### References to various things

- Add a trailing slash to directory references to distinguish from regular files.

> - [Link to Silverstripe API docs](#links-to-api-documentation) for the first reference to a Silverstripe PHP class on a page. After that, surrond the class name with backticks.

Reviewer: ```suggestion
- [Link to Silverstripe API docs](#links-to-api-documentation) for the first reference to a Silverstripe PHP class on a page. For subsequent references, use just the class name surrounded with backticks.
```

Author: Done

- Use `example.com` as the domain of sample URLs and `example.org` and `example.net` when additional domains are required. All of these domains are [reserved by the IANA](https://tools.ietf.org/html/rfc2606#section-3).

> ### References to Silverstripe

Reviewer: We don't need a specific heading for this - just fold this into the "references to various things" section

Author: Done

- Spell "Silverstripe" with an uppercase 'S' at the beginning and a lowercase 's' in the middle.

> - Refer to the company as "Silverstripe Ltd."

Reviewer: ```suggestion
- Refer to the company as "Silverstripe."
```
This is how all the documentation now refers to the company. It used to be a mix of "SilverStripe", "Silverstripe", and "Silverstripe ltd" but at some point we made the decision to drop the ltd. If we want to change that we need to update all the docs, and probably references elsewhere on .org as well.

Author: Done

> - Use "Silverstripe CMS" to refer the content management system.

Reviewer: ```suggestion
- Use "Silverstripe CMS" to refer to the Silverstripe CMS open source project as a whole
```
"the content management system" isn't wide-spanding enough IMO. The original language was clearer.

Author: Done

### Markdown linting

- Files that are within a directory specified in repository's `.doclintrc` file will be automatically linted in CI using [silverstripe/documentation-lint](https://github.com/silverstripe/documentation-lint/).

> - Many of the linting rules are not mentioned in this writing style guide, though all linting rules will need to be adherred to in order for your documentation pull-request to be merged. The failed linting CI build will clearly highlight what needs to be fixed.

Reviewer: We definitely don't need the technical details about a `.doclintrc` file - IMO we don't need this section at all but if you want to keep it, make the first point just something like "Documentation will be automatically linted in CI"

Author: Done
